amplicon_stats.c

```diff
@@ -0,0 +1,943 @@
+DEALINGS IN THE SOFTWARE.  */
+
+/*
+ * This tool is designed to give "samtools stats" style output, but dedicated
+ * to small amplicon sequencing projects.  It gathers stats on the
+ * distribution of reads across amplicons.
+ */
+
+/*
+ * TODO:
+ * - Cope with arbitrary sizes of reference and numbers of amplicons
+ * - Cope with multiple references.  What do we do here?  Just request one?
+ * - Permit regions rather than consuming whole file (maybe solves above).
+ * - Detect double-length products (or primer pool hopping).
+ * - eg 31_LEFT to 33_RIGHT, or even 31_LEFT to 32_RIGHT.
+ *   Binning via TLEN field will suffice.
+ */
+
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
+#include <getopt.h>
+#include <unistd.h>
+#include <math.h>
+
+#include <htslib/sam.h>
+#include <htslib/hfile.h>
+
+#include "samtools.h"
+#include "sam_opts.h"
+#include "bam_ampliconclip.h"
+
+#ifndef MIN
+#define MIN(a,b) ((a)<(b)?(a):(b))
+#endif
+
+#ifndef MAX
+#define MAX(a,b) ((a)>(b)?(a):(b))
+#endif
+
+#ifndef ABS
+#define ABS(a) ((a)>=0?(a):-(a))
+#endif
+
+#define MAX_AMP_LEN 1000   // Maximum length of any single amplicon
+#define MAX_PRIMER_PER_AMPLICON 4  // Max primers per LEFT/RIGHT
+#define MAX_DEPTH 5        // Number of different depths permitted
+#define MAX_TLEN 2000      // TLEN distribution size
+
+typedef struct {
+    sam_global_args ga;
+    uint32_t flag_require;
+    uint32_t flag_filter;
+    bed_pair_list_t sites;
+    int max_delta; // Used for matching read to amplicon primer loc
+    int min_depth[MAX_DEPTH]; // Used for coverage; must be >= min_depth deep
+    int use_sample_name;
+    int max_amp;
+    int64_t max_len;
+    int tlen_adj; // Adjust tlen by this amount, due to clip but no fixmate
+    FILE *out_fp;
+} astats_args_t;
+
+// We can have multiple primers for LEFT / RIGHT, so this
+// permits detection by any compatible combination.
+typedef struct {
+    int64_t left[MAX_PRIMER_PER_AMPLICON];
+    int nleft;
+    int64_t right[MAX_PRIMER_PER_AMPLICON];
+    int nright;
+    int64_t max_left, min_right; // inner dimensions
+    int64_t min_left, max_right; // outer dimensions
+} amplicon_t;
+
+// Map positions to amplicon numbers.
+static int *pos2start = NULL;
+static int *pos2end = NULL;
+static int64_t pos_lookup_len = 0;
+
+// Lookup table to go from position to amplicon based on
+// read start / end.
+//
+// NB: Could do bed2amplicon and this code as bed2pos() in a single step.
+static int initialise_amp_pos_lookup(astats_args_t *args,
+                                     amplicon_t *amp, int namp,
+                                     int64_t max_len) {
+    int64_t i, j;
+
+    if (!pos_lookup_len) {
+        pos_lookup_len = max_len;
+        if (!(pos2start = calloc(max_len+1, sizeof(*pos2start))))
+            return -1;
+        if (!(pos2end   = calloc(max_len+1, sizeof(*pos2start)))) {
+            free(pos2start);
```

Comment:
Just noticed a small bug.  If `pos2end` fails to allocate, `pos2start` will be double freed by `free_amp_pos_lookup()` if it's also done here.

Reply:
Good spot.  Fixed.